### DISCUSSION TITLE
Use circuitikz package in pytikz

### DISCUSSION THREAD

**Discussion by Xuanyiyiren** [allefeld/pytikz#11] [Ideas]
In my work and study, I often use another package [circuitikz](https://texdoc.org/serve/circuitikzmanual.pdf/0) to draw some circuit diagrams with electronic component such as resistors, transistors and some logic gates. So I want to add this function to pytikz.

But I am confused about how pytikz generate pdf from tikz code in `Picture.code()`.  I think this is achieved by `subprocess` , right? So if the document code can give an appropriate output in $\LaTeX$, so do pytikz. But things seem not so simple.

I am familiar with circuitikz.So [I forked your repository](https://github.com/Xuanyiyiren/pytikz) and make some change on the `__init__.py` to generator appropriate code using package 'circuitikz'. 

And then I run it in python. All the old functions run well and the same as before. But the program seems to be stuck in an infinite loop when I try to use circuitikz package. You can see it in the last cell in [mytest-circuit.ipynb](https://github.com/Xuanyiyiren/pytikz/blob/master/mytest-circuitikz.ipynb). The `document_code` can give right output in overleaf, you can see it [here](https://www.overleaf.com/read/wsrzkvcqrmbp#924b23).

However, I have not received professional training in computer science. I am just an undergraduate student majoring in physics. It seems that I do not have the enough ability and energy to solve this. I would appreciate your help with this problem.

Use circuitikz is much more complicated then just tikz. So this ideal may not be as simple as I think. Even this ideal is not necessary enough because there are many other ways to create circuit diagrams. 

Nevertheless, the circuitikz package on $\LaTeX$ can combine circuit diagrams with latex equations. And more importantly, circuitikz has a very complete and professional library of electronic components that can be freely edited.(Many circuit diagram drawers cannot give a 3-input AND gate or even n-input AND gate. This can be easily achieved by circuitikz. The example can be found in the 166th page of its [documentation](https://texdoc.org/serve/circuitikzmanual.pdf/0).) This is generally only possible with electronic circuit simulation software, which does not have the function of visualizing beautiful circuit diagrams.

Meanwhile, I am not a native English speaker, so please let me know if anything I say is unclear.

**Comment by allefeld** [comment]
It seems `circuitikz` introduces a new type of environment similar to `tikzpicture`, and you're right, that needs a modification of the Python code. You seem to have modified the `Picture` class, but a better approach would be to define a new class `Circuit` which inherits from `Picture`, changing the name of the environment in the produced LaTeX code to `circuitikz` and maybe also automating `usepackage('circuitikz')` – but your approach should work, too.

You're correct that LaTeX is run by `subprocess` (see [`__init__.py` line 1166](https://github.com/Xuanyiyiren/pytikz/blob/606d96c2a3583303d6a44bd57f6555209e98e290/tikz/__init__.py#L1166)), and if the LaTeX code produced by the package works otherwise, it should work in a notebook, too. Unfortunately, from the materials you shared I can't see what might have gone wrong, and right now I don't have the time to install and try your modified version. You're saying there is an infinite loop; I think it's more likely that the LaTeX process failed but the error was not captured – if so, that is something that should be captured by my code.

One thing you could check is whether `circuitikz` is compatible with XeLaTeX, which is used by default (see [`tikz.cfg.latex`](https://allefeld.github.io/pytikz/tikz/#tikz.cfg.latex)). If not, try `tikz.cfg.latex = 'pdflatex'`.

What you could also try to do is run code in a Python script instead of a notebook. Copy the code from the first and the last cell into a file `mytest-circuitikz.py`. Hopefully there would be more visible output, which allows us to figure out what is going wrong.
